### PR TITLE
infoschema: adapt TableByName for query of memory table

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/50305

Problem Summary:

Read memory tables from `information_schema` is slow when `tidb_schema_cache_size` is non-zero.

### What changed and how does it work?

- Add `TableSchemaSelector` to help return the table/schema names specified in predicate.
- Adapt `TableByName` for memory table `information_schema.tables`.
- Adapt `FindTableByPartitionID` for `information_schema.partitions`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  Before this PR:
  ```
  -- first load
  mysql> select * from  information_schema.tables where table_schema = 'test' and table_name = 't6613';
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  | TABLE_CATALOG | TABLE_SCHEMA | TABLE_NAME | TABLE_TYPE | ENGINE | VERSION | ROW_FORMAT | TABLE_ROWS | AVG_ROW_LENGTH | DATA_LENGTH | MAX_DATA_LENGTH | INDEX_LENGTH | DATA_FREE | AUTO_INCREMENT | CREATE_TIME         | UPDATE_TIME | CHECK_TIME | TABLE_COLLATION | CHECKSUM | CREATE_OPTIONS | TABLE_COMMENT | TIDB_TABLE_ID | TIDB_ROW_ID_SHARDING_INFO | TIDB_PK_TYPE | TIDB_PLACEMENT_POLICY_NAME |
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  | def           | test         | t6613      | BASE TABLE | InnoDB |      10 | Compact    |          2 |              8 |          16 |               0 |           16 |         0 |              5 | 2024-07-25 21:10:37 | NULL        | NULL       | utf8mb4_bin     |     NULL |                |               |          8063 | NOT_SHARDED               | NONCLUSTERED | NULL                       |
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  1 row in set (22.68 sec)
  
  
  -- subsequent load
  mysql> select * from  information_schema.tables where table_schema = 'test' and table_name = 't6613';
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  | TABLE_CATALOG | TABLE_SCHEMA | TABLE_NAME | TABLE_TYPE | ENGINE | VERSION | ROW_FORMAT | TABLE_ROWS | AVG_ROW_LENGTH | DATA_LENGTH | MAX_DATA_LENGTH | INDEX_LENGTH | DATA_FREE | AUTO_INCREMENT | CREATE_TIME         | UPDATE_TIME | CHECK_TIME | TABLE_COLLATION | CHECKSUM | CREATE_OPTIONS | TABLE_COMMENT | TIDB_TABLE_ID | TIDB_ROW_ID_SHARDING_INFO | TIDB_PK_TYPE | TIDB_PLACEMENT_POLICY_NAME |
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  | def           | test         | t6613      | BASE TABLE | InnoDB |      10 | Compact    |          2 |              8 |          16 |               0 |           16 |         0 |              5 | 2024-07-25 21:10:37 | NULL        | NULL       | utf8mb4_bin     |     NULL |                |               |          8063 | NOT_SHARDED               | NONCLUSTERED | NULL                       |
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  1 row in set (2.51 sec)
  ```
  After this PR:
  ```
  mysql> select * from  information_schema.tables where table_schema = 'test' and table_name = 't6613';
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  | TABLE_CATALOG | TABLE_SCHEMA | TABLE_NAME | TABLE_TYPE | ENGINE | VERSION | ROW_FORMAT | TABLE_ROWS | AVG_ROW_LENGTH | DATA_LENGTH | MAX_DATA_LENGTH | INDEX_LENGTH | DATA_FREE | AUTO_INCREMENT | CREATE_TIME         | UPDATE_TIME | CHECK_TIME | TABLE_COLLATION | CHECKSUM | CREATE_OPTIONS | TABLE_COMMENT | TIDB_TABLE_ID | TIDB_ROW_ID_SHARDING_INFO | TIDB_PK_TYPE | TIDB_PLACEMENT_POLICY_NAME |
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  | def           | test         | t6613      | BASE TABLE | InnoDB |      10 | Compact    |          2 |              8 |          16 |               0 |           16 |         0 |              1 | 2024-07-25 21:10:37 | NULL        | NULL       | utf8mb4_bin     |     NULL |                |               |          8063 | NOT_SHARDED               | NONCLUSTERED | NULL                       |
  +---------------+--------------+------------+------------+--------+---------+------------+------------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------------+---------------+---------------------------+--------------+----------------------------+
  1 row in set (0.00 sec)
  ```
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
@tangenta 

```
[2024/07/25 14:54:35.357 +00:00] [INFO] [schema_version.go:381] ["wait latest schema version changed(get the metadata lock if tidb_enable_metadata_lock is true)"] [category=ddl] [ver=52] ["take time"=2.815636ms] [job="ID:108, Type:create table, State:done, SchemaState:public, SchemaID:2, TableID:104, RowCount:0, ArgLen:2, start time: 2024-07-25 14:54:35.351 +0000 UTC, Err:<nil>, ErrCount:0, SnapshotVersion:0, LocalMode: false"]
[2024/07/25 14:54:35.359 +00:00] [INFO] [ddl_worker.go:940] ["finish DDL job"] [worker="worker 22, tp general"] [category=ddl] [jobID=108] [conn=3] [job="ID:108, Type:create table, State:synced, SchemaState:public, SchemaID:2, TableID:104, RowCount:0, ArgLen:0, start time: 2024-07-25 14:54:35.351 +0000 UTC, Err:<nil>, ErrCount:0, SnapshotVersion:0, LocalMode: false"]
[2024/07/25 14:54:35.361 +00:00] [INFO] [ddl.go:1271] ["DDL job is finished"] [category=ddl] [jobID=108]
    testkit.go:274: 
        	Error Trace:	pkg/testkit/testkit.go:274
        	            				pkg/testkit/testkit.go:199
        	            				pkg/testkit/testkit.go:174
        	            				pkg/planner/core/memtable_predicate_extractor_test.go:1679
        	Error:      	Received unexpected error:
        	            	[schema:1146]Table 'INFORMATION_SCHEMA.p' doesn't exist
        	            	github.com/pingcap/errors.AddStack
        	            		external/com_github_pingcap_errors/errors.go:178
        	            	github.com/pingcap/errors.Trace
        	            		external/com_github_pingcap_errors/juju_adaptor.go:15
        	            	github.com/pingcap/tidb/pkg/executor.getAllTableInfos
        	            		pkg/executor/infoschema_reader.go:588
        	            	github.com/pingcap/tidb/pkg/executor.(*memtableRetriever).setDataFromTables
        	            		pkg/executor/infoschema_reader.go:631
        	            	github.com/pingcap/tidb/pkg/executor.(*memtableRetriever).retrieve
        	            		pkg/executor/infoschema_reader.go:133
        	            	github.com/pingcap/tidb/pkg/executor.(*MemTableReaderExec).Next
        	            		pkg/executor/memtable_reader.go:120
        	            	github.com/pingcap/tidb/pkg/executor/internal/exec.Next
        	            		pkg/executor/internal/exec/executor.go:451
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).next
        	            		pkg/executor/adapter.go:1249
        	            	github.com/pingcap/tidb/pkg/executor.(*recordSet).Next
        	            		pkg/executor/adapter.go:169
        	            	github.com/pingcap/tidb/pkg/session.GetRows4Test
        	            		pkg/session/tidb.go:351
        	            	github.com/pingcap/tidb/pkg/session.ResultSetToStringSlice
        	            		pkg/session/tidb.go:369
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).ResultSetToResultWithCtx
        	            		pkg/testkit/testkit.go:273
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustQueryWithContext
        	            		pkg/testkit/testkit.go:199
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustQuery
        	            		pkg/testkit/testkit.go:174
        	            	github.com/pingcap/tidb/pkg/planner/core_test.TestTikvRegionStatusExtractor
        	            		pkg/planner/core/memtable_predicate_extractor_test.go:1679
        	            	testing.tRunner
        	            		GOROOT/src/testing/testing.go:1595
        	            	runtime.goexit
        	            		src/runtime/asm_amd64.s:1650
        	Test:       	TestTikvRegionStatusExtractor
        	Messages:   	sql:select * from information_schema.tables where table_name = 'p', args:[]
[2024/07/25 14:54:35.362 +00:00] [INFO] [mock.go:101] ["owner manager retire owner"] [ownerKey=/tidb/ddl/fg/owner] [ID=536f52c9-7647-4fea-bf30-15b953339b6d]
[2024/07/25 14:54:35.362 +00:00] [INFO] [job_table.go:268] ["scheduleLoop quit due to context canceled"] [category=ddl]
```

**Review by tiancaiamao — APPROVED** [review]

File: tests/integrationtest/r/infoschema/infoschema.result [hunk 1]
```diff
@@ -115,6 +115,7 @@ Projection_4	8000.00	root		Column#5, Column#10
   └─MemTableScan_6	10000.00	root	table:TABLES	table_schema:["infoschema__infoschema"]
 select engine, DATA_LENGTH from information_schema.tables where lower(table_name) = 't5' and  upper(table_schema) = 'INFOSCHEMA__INFOSCHEMA';
 engine	DATA_LENGTH
+InnoDB	8
```
Comment (tangenta) [inline tests/integrationtest/r/infoschema/infoschema.result:118]: There is a bug in previous impl.

File: tests/integrationtest/t/infoschema/infoschema.test [hunk 2]
```diff
@@ -58,3 +58,14 @@ select engine, DATA_LENGTH from information_schema.tables where (table_name ='t4
 explain select engine, DATA_LENGTH from information_schema.tables where table_name ='t4' and upper(table_name) ='T4' and table_schema = 'infoschema__infoschema';
 select engine, DATA_LENGTH from information_schema.tables where table_name ='t4' and upper(table_name) ='T4' and table_schema = 'infoschema__infoschema';
 
+# TestPartitionsColumn
+create table pt1(a int primary key, b int) partition by hash(a) partitions 4;
+create table pt2(a int primary key, b int) partition by hash(a) partitions 4;
+-- sorted_result
```
Comment (lance6716) [inline tests/integrationtest/t/infoschema/infoschema.test:64]: I'm not familiar with these logic. Can these test make sure the execution does not query full data in TiKV?

Comment (tangenta) [reply tests/integrationtest/t/infoschema/infoschema.test:64]: No. It only make sure the result is correct. Let me add a manual test about `information_schema.partitions`.

File: pkg/planner/core/memtable_predicate_extractor.go [hunk 3]
```diff
@@ -1825,7 +1826,8 @@ func (e *InfoSchemaTablesExtractor) Extract(ctx base.PlanContext,
 	predicates []expression.Expression,
 ) (remained []expression.Expression) {
 	var resultSet, resultSet1 set.StringSet
-	e.colNames = []string{"table_schema", "constraint_schema", "table_name", "constraint_name", "sequence_schema", "sequence_name", "partition_name", "schema_name", "index_name"}
+	e.colNames = []string{"table_schema", "constraint_schema", "table_name", "constraint_name",
```
Comment (lance6716) [inline pkg/planner/core/memtable_predicate_extractor.go:1829]: we don't need to use a struct field `e.colNames`. We can use a global variable instead

File: pkg/executor/infoschema_reader.go [hunk 4]
```diff
@@ -563,151 +587,275 @@ func (e *memtableRetriever) updateStatsCacheIfNeed() bool {
...
-	var rows [][]types.Datum
-	createTimeTp := mysql.TypeDatetime
-	loc := sctx.GetSessionVars().TimeZone
-	if loc == nil {
-		loc = time.Local
+func getMatchSchemas(
+	extractor base.MemTablePredicateExtractor,
+	is infoschema.InfoSchema,
+) []model.CIStr {
+	ex := extractor.(plannercore.TableSchemaSelector)
```
Comment (lance6716) [inline pkg/executor/infoschema_reader.go:594]:
```suggestion
	ex , ok := extractor.(plannercore.TableSchemaSelector)
```

will panic if not `ok`. If you have confident to make sure `TableSchemaSelector` is always implemented it's no need to check `if ex != nil` in line 595?

File: pkg/executor/infoschema_reader.go [hunk 5]
```diff
@@ -563,151 +587,275 @@ func (e *memtableRetriever) updateStatsCacheIfNeed() bool {
...
+}
+
+func getMatchTableInfosForPartitions(
+	ctx context.Context,
+	extractor base.MemTablePredicateExtractor,
+	schema model.CIStr,
+	is infoschema.InfoSchema,
+) ([]*model.TableInfo, error) {
+	ex := extractor.(plannercore.TableSchemaSelector)
+	if ex == nil || !ex.HasTables() {
```
Comment (lance6716) [inline pkg/executor/infoschema_reader.go:620]: ditto

File: pkg/executor/infoschema_reader.go [hunk 6]
```diff
@@ -563,151 +587,275 @@ func (e *memtableRetriever) updateStatsCacheIfNeed() bool {
...
+}
+
+func getMatchTableInfos(
+	ctx context.Context,
+	extractor base.MemTablePredicateExtractor,
+	schema model.CIStr,
+	is infoschema.InfoSchema,
+) ([]*model.TableInfo, error) {
+	ex := extractor.(plannercore.TableSchemaSelector)
+	if ex == nil || !ex.HasTables() {
```
Comment (lance6716) [inline pkg/executor/infoschema_reader.go:656]: ditto

File: pkg/executor/infoschema_reader.go [hunk 7]
```diff
@@ -563,151 +587,275 @@ func (e *memtableRetriever) updateStatsCacheIfNeed() bool {
...
-			createTime := types.NewTime(types.FromGoTime(table.GetUpdateTime().In(loc)), createTimeTp, types.DefaultFsp)
+			return nil, errors.Trace(err)
+		}
+		tables = append(tables, tbl.Meta())
+	}
+	return deduplicateTableInfos(tables), nil
+}
 
-			createOptions := ""
+func deduplicateTableInfos(tables []*model.TableInfo) []*model.TableInfo {
```
Comment (lance6716) [inline pkg/executor/infoschema_reader.go:689]: maybe at caller directly use `map[int64]*model.TableInfo` and use `maps.Values()` to get the slice to return.

**Review by lance6716 — COMMENTED** [review]
rest lgtm

**Review by lance6716 — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

File: pkg/executor/infoschema_reader.go [hunk 8]
```diff
@@ -112,31 +114,42 @@ func (e *memtableRetriever) retrieve(ctx context.Context, sctx sessionctx.Contex
...
+			slices.SortFunc(dbs, func(a, b model.CIStr) int {
+				return strings.Compare(a.L, b.L)
+			})
+			return dbs
+		}
+
 		var err error
 		switch e.table.Name.O {
 		case infoschema.TableSchemata:
+			dbs := getAllSchemas()
```
Comment (AilinKid) [inline pkg/executor/infoschema_reader.go:129]: why not eval them once before switch case

Comment (tangenta) [reply pkg/executor/infoschema_reader.go:129]: Because some of the cases don't need this.

**Pull request merged**